XMonad/Prompt/DotDesktop.hs

```diff
@@ -0,0 +1,146 @@
+cmdFilter ('%':'%':xs) = '%' : cmdFilter xs
+cmdFilter (x:xs) = x : cmdFilter xs
+cmdFilter "" = ""
+
+convertExceptionToString :: Exception e => IO (Either e a) -> IO (Either String a)
+convertExceptionToString = fmap convertExceptionToStringHelper
+
+convertExceptionToStringHelper :: Exception e => Either e a -> Either String a
+convertExceptionToStringHelper = either (Left . convertExceptionToStringHelperHelper) Right
+
+convertExceptionToStringHelperHelper :: Exception e => e -> String
+convertExceptionToStringHelperHelper = show :: Exception e => e -> String
+
+doReadFileLBS :: String -> ExceptT String IO String
+doReadFileLBS = ExceptT . convertExceptionToString . try @IOError . readFile
+
+getVal :: String -> String -> M.Map String String -> Either String String
+getVal msg k kvmap = maybeToEither msg $ M.lookup k kvmap
+
+maybeToEither :: b -> Maybe a -> Either b a
+maybeToEither _ (Just a) = Right a
+maybeToEither b Nothing = Left b
+
+doParseFile :: String -> ExceptT String IO DotDesktopApp
+doParseFile filePath = do
+  content <- doReadFileLBS filePath
+  parsed <- liftEither $ runDotDesktopParser content
+  let kvMaybe = snd <$> listToMaybe (rights parsed)
+  keyVals <- liftEither $
+    maybe
+      (Left $ "Parse Resulted in no KeyVals in file " ++ filePath)
+      Right
+      kvMaybe
+  let errMsg = "Unable to find Name in file " ++ filePath
+  nom <- liftEither $ getVal errMsg "Name" keyVals
+  exc <- liftEither $ getVal errMsg "Exec" keyVals
+  typ <- liftEither $ getVal errMsg "Type" keyVals
+  return DotDesktopApp { fileName = filePath
+                       , name = nom
+                       , type_ = typ
+                       , exec = exc
+                       , cmd = (trimWhitespace . cmdFilter) exc
+                       }
+
+data DotDesktopApp = DotDesktopApp { fileName :: String
+                             , name :: String
+                             , type_ :: String
+                             , exec :: String
+                             , cmd :: String
+                             } deriving Show
+
+getXdgDataHome :: IO (Maybe FilePath)
+getXdgDataHome = do
+  envXdgDataHome <- envXdgDataHomeIO
+  defaultXdgDataHome <- defaultXdgDataHomeIO
+  return $ envXdgDataHome <|> defaultXdgDataHome
+  where
+    defaultXdgDataHomeIO = lookupEnv "HOME" <&> fmap (</> ".local" </> "share")
+    envXdgDataHomeIO = lookupEnv "XDG_DATA_HOME"
+
+getXdgDataDirs :: IO [FilePath]
+getXdgDataDirs =
+  fromMaybe defaultXdgDataDirs <$> envXdgDataDirsIO
+  where
+    defaultXdgDataDirs = split ':' "/usr/local/share:/usr/share"
+    envXdgDataDirsIO = lookupEnv "XDG_DATA_DIRS" <&> (<&> split ':')
+
+getAppFolders :: IO [FilePath]
+getAppFolders = do
+  xdgDataHome <- maybeToList <$> getXdgDataHome
+  xdgDataDirs <- getXdgDataDirs
+  let possibleAppDirs = xdgDataHome ++ xdgDataDirs <&> (</> "applications")
+  filterM doesDirectoryExist possibleAppDirs
+
+getDirContents :: FilePath -> ExceptT String IO [FilePath]
```

Comment:
I don't think ExceptT is really necessary here, since all you do with is is run it immediately after calling the function.  Why not just return an `IO (Either String [FilePath])` instead?

Reply:
Agreed.  I pulled on this and ended up removing all ExceptT